runtime/vm/criuhelpers.cpp

```diff
@@ -103,4 +105,259 @@ isCheckpointAllowed(J9VMThread *currentThread)
+{
+	J9Pool *hookRecords = vm->checkpointState.hookRecords;
+	if (NULL != hookRecords) {
+		pool_state walkState = {0};
+		J9InternalHookRecord *hookRecord = (J9InternalHookRecord*)pool_startDo(hookRecords, &walkState);
+		while (NULL != hookRecord) {
+			pool_kill(hookRecord->instanceObjects);
+			hookRecord->instanceObjects = NULL;
+			hookRecord = (J9InternalHookRecord*)pool_nextDo(&walkState);
+		}
+
+		if (vm->checkpointState.isNonPortableRestoreMode) {
+			/* No more checkpoint, cleanup hook records */
+			pool_kill(vm->checkpointState.hookRecords);
+			vm->checkpointState.hookRecords = NULL;
+		}
+	}
+}
+
+/**
+ * After initializeCriuHooks() and before invoking JVM criu hooks, j9mm_iterate_all_objects iterates the heap,
+ * for each object discovered, this will be called to iterate CRIU internal hook records, and fills in all instances of
+ * instanceType and its subclasses if specified.
+ *
+ * @param[in] vm J9JavaVM
+ * @param[in] object an instance to be checked and filled in if it is the instanceType specified
+ *
+ * @return BOOLEAN TRUE if no error, otherwise FALSE
+ */
+static BOOLEAN
+fillinHookRecords(J9JavaVM *vm, j9object_t object)
+{
+	J9Pool *hookRecords = vm->checkpointState.hookRecords;
+	pool_state walkState = {0};
+	BOOLEAN result = TRUE;
+	J9VMThread *currentThread = vm->mainThread;
+
+	J9InternalHookRecord *hookRecord = (J9InternalHookRecord*)pool_startDo(hookRecords, &walkState);
+	while (NULL != hookRecord) {
+		J9Class *clazz = J9OBJECT_CLAZZ_VM(vm, object);
+		if ((clazz == hookRecord->instanceType)
+			|| (hookRecord->includeSubClass && isSameOrSuperClassOf(hookRecord->instanceType, clazz))
+		) {
+			if (NULL == hookRecord->instanceObjects) {
+				hookRecord->instanceObjects = pool_new(sizeof(J9Object), 0, 0, 0, J9_GET_CALLSITE(), OMRMEM_CATEGORY_VM, POOL_FOR_PORT(vm->portLibrary));
+				if (NULL == hookRecord->instanceObjects) {
+					setNativeOutOfMemoryError(currentThread, 0, 0);
+					result = FALSE;
+					break;
+				}
+			}
+			J9Object *objectRecord = (J9Object*)pool_newElement(hookRecord->instanceObjects);
+			if (NULL == objectRecord) {
+				result = FALSE;
+				setNativeOutOfMemoryError(currentThread, 0, 0);
+			} else {
+				objectRecord->clazz = (j9objectclass_t)object;
+			}
+		}
+
+		hookRecord = (J9InternalHookRecord*)pool_nextDo(&walkState);
+	}
+
+	return result;
+}
+
+static jvmtiIterationControl
+objectIteratorCallback(J9JavaVM *vm, J9MM_IterateObjectDescriptor *objectDesc, void *userData)
+{
+	j9object_t object = objectDesc->object;
+
+	fillinHookRecords(vm, object);
+
+	return JVMTI_ITERATION_CONTINUE;
+}
+
+BOOLEAN
+runInternalJVMCheckpointHooks(J9JavaVM *vm)
+{
+	J9VMThread *currentThread = vm->mainThread;
+	J9Pool *hookRecords = vm->checkpointState.hookRecords;
+	pool_state walkState = {0};
+	BOOLEAN result = TRUE;
+
+	Trc_VM_criu_runCheckpointHooks_Entry(currentThread);
+
+	/* Iterate heap objects to prepare internal hooks at checkpoint */
+	acquireExclusiveVMAccess(currentThread);
```

Comment:
same here, lets acquire in the jcl native so its more visible to readers

Reply:
we can add a comment as well in the native saying that the internalhooks hold hold to live object refs, so GC cannot be allowed to run.

Reply:
Moved it into `Java_org_eclipse_openj9_criu_CRIUSupport_checkpointJVMImpl`.
Added comments 
_ExclusiveVMAccess is required since J9InternalHookRecord hold live object references
and GC are not allowed while running these hook functions._